test/jdk/javax/swing/JTextArea/4514331/TabShiftsFocusToNextComponent.java

```diff
@@ -0,0 +1,140 @@
+import java.util.stream.Collectors;
+import javax.swing.JButton;
+import javax.swing.JFrame;
+import javax.swing.JPanel;
+import javax.swing.JTextArea;
+import javax.swing.SwingUtilities;
+import javax.swing.UIManager;
+
+import static javax.swing.UIManager.getInstalledLookAndFeels;
+
+/*
+ * @test
+ * @key headful
+ * @bug 4514331
+ * @summary Check whether pressing <Tab> key always shift focus to next component,
+ *          even though the current focus is in JTextArea and some text is already selected.
+ * @run main TabShiftsFocusToNextComponent
+ */
+public class TabShiftsFocusToNextComponent {
+
+    private static JFrame frame;
+    private static JTextArea editor;
+    private static Robot robot;
+    private static boolean passed = false;
+
+    public static void main(String[] s) throws Exception {
+        runTest();
+    }
+
+    public static void runTest() throws Exception {
+        robot = new Robot();
+        robot.setAutoWaitForIdle(true);
+        robot.setAutoDelay(200);
+        List<String> lafs = Arrays.stream(getInstalledLookAndFeels())
+                                  .map(UIManager.LookAndFeelInfo::getClassName)
+                                  .collect(Collectors.toList());
+        for (final String laf : lafs) {
+            try {
+                SwingUtilities.invokeAndWait(() -> frame = new JFrame());
+                robot.waitForIdle();
+                AtomicReference<Point> editorLoc = new AtomicReference<Point>();
+                SwingUtilities.invokeAndWait(() -> {
+                    setLookAndFeel(laf);
+                    createUI();
+                    editorLoc.set(editor.getLocationOnScreen());
+                });
+                passed = false;
+
+                final int x = editorLoc.get().x;
+                final int y = editorLoc.get().y;
+                robot.mouseMove(x, y);
+                robot.mousePress(InputEvent.BUTTON1_DOWN_MASK);
+                robot.mouseMove(x + 20, y);
+                robot.mouseRelease(InputEvent.BUTTON1_DOWN_MASK);
+                robot.keyPress(KeyEvent.VK_TAB);
+                robot.keyRelease(KeyEvent.VK_TAB);
+                if (passed) {
+                    System.out.println(" Test passed for " + laf);
+                } else {
+                    throw new RuntimeException("Test failed for " + laf);
+                }
+            } finally {
+                SwingUtilities.invokeAndWait(TabShiftsFocusToNextComponent::disposeFrame);
+            }
+        }
+    }
+
+
+    private static void createUI() {
+        JPanel panel = new JPanel();
+        editor = new JTextArea("I am a JTextArea");
+        editor.setEditable(false);
+        panel.add(editor);
+        JButton button = new JButton("Button");
+        panel.add(button);
+        button.addFocusListener(new FocusAdapter() {
+            @Override
+            public void focusGained(FocusEvent e) {
+                passed = true;
```

Comment:
the "passed" var is set on EDT and then used on the main thread, I suggest marking it as volatile.